### DISCUSSION TITLE
Multi material single extruder

### DISCUSSION THREAD

**Discussion by Kalorado303** [bassamanator/Sovol-SV06-firmware#57] [Q&A]
Hello Bassamanator, I recently installed kipper on my sv06 using your girhub instructions. Thank you so much as it had been working great! I just had a few questions. I am trying to do multi material namely petg on tpu. I have set up two extruders and clicked single nozzle in super slicer then assigned each material to an extruder. I also added an m600 at my needed layer height. When I went to print I got three beeps from the printer indicating it read the m600 at the correct layer but then it just parked the head and I had an error in mainsail saying extruder1 was not recognized. Do I need to add code for a secondary extruded into klipper to get this to work? Also when I tried the m600 macro manually it seems to pause the print and park the print head then show a ready state on the lcd and nothing else. Is that the intended behavior or am I missing something?  Again thank you so much for your contribution as it has made this hobby much more fun.

**Comment by bassamanator** [comment]
>Also when I tried the m600 macro manually it seems to pause the print and park the print head then show a ready state on the lcd and nothing else. Is that the intended behavior or am I missing something?

Please see [this section](https://github.com/bassamanator/Sovol-SV06-firmware#what-happens-when-i-put-in-m600colour-change-at-a-certain-layer) for M600 behaviour. Read entire FAQ section while there.

I've never done a multi-material print, but it would probably go as follows:
- Add M600 at the desired layer. You could even use a pause I suppose.
- Set new material temperature in mainsail/fluidd.
- Load new filament.
- Hit resume.

It's likely that Klipper will revert to using the first material temperature when the print resumes, so you'll likely have to set the new material temperature again.

Give that a shot and let us know what happened.

**Comment by bassamanator** [comment]
Also, I recommend trying the above on a 'fake' print. Just use a tiny cube, and add the colour change on layer 2. Don't bother actually changing materials either. You're just trying to ascertain the behaviour.

**Comment by Kalorado303** [comment]
I will give that a shot. One quick question as I am new to Klipper. Do I have to pull the macro code into the printer config file? Or are they fine in the misc config file where they are. If that is the case, is that where I would add other macros?

**Comment by bassamanator** [comment]
You only need follow the instructions. If you've done that, you're done and ready to go.

**Comment by Kalorado303** [comment]
I tried the M600 based on your instructions and while it works a few things of note. It appears as though the resume option in the lcd under "octoprint" only gives an "action:resume" line in the console but does not do anything I had to initiate resume from mainsail. Is there a way to get this button to work so everything is doable at the printer?

Also the biggest issue and the reason I would like to get klipper to recognize multiple materials via super slicer is due to the vastly different settings ie. flow from TPU to PETG not just the temperature change. 

And also is there a way to include a purge line at the start of resume?

**Reply by bassamanator** [reply]
I haven't used Octoprint in years, so I'm not sure what would need to happen there.

Good point about the other settings, they're just as important as the temperature. If this can be done, then it's already been done, and you should search for that solution. I'm curious about this myself now, though I'm pessimistic about whether this can be done.

**Reply by Kalorado303** [reply]
Sorry I am using mainsail and have set it up per your instructions I only meant the resume option in the lcd/scrollwheel is labeled octoprint.

**Reply by Kalorado303** [reply]
I assigned two extruders in SS with one nozzle and inserted a change extruder g-code in the slicer. If you tell SS what filament each extruder has it appears to adjust all parameters after the change. The only other thing I did was add M600 to SS tool change and filament change G-code in filament settings.

Everything seems to work fine but it appears that I am getting an M600 or pause ( hard to tell) after the initial purge line and before print begins so I have to manually resume the print before it actually starts and I am not sure why. 

That and the aforementioned resume print not working through the lcd/knob.

**Reply by Kalorado303** [reply]
I checked my g code and it appears that SS is calling for a tool change for the initial extruder and since I added M600 in the slicer tool change profile it is executing an M600 command and pausing the print. Unless you know how to correct this I am going to remove the tool change custom g code from SS and manually add a color change at the same time as the tool/filament change. Here is the g code I am referencing.
M73 P0 R94
;TYPE:Custom
M104 S0 ; Stops PS/SS from sending temp waits separately
M140 S0
PRINT_START BED=70 HOTEND=220 CHAMBER=0
PURGE_LINE

;; If you are passing variables to PRINT_START, comment out above line and uncomment the last 3 lines instead. (I already did this)
;; Guide: https://github.com/AndrewEllis93/Print-Tuning-Guide/blob/main/articles/passing_slicer_variables.md
;; Make sure the variable names match (for example if you use "EXTRUDER" instead of "HOTEND")
M107
G21 ; set units to millimeters
G90 ; use absolute coordinates
M83 ; use relative distances for extrusion
M600
;_TOOLCHANGE 0
M900 K0
M107
;LAYER_CHANGE
;Z:0.26
;HEIGHT:0.26

**Comment by Kalorado303** [comment]
Disregard my previous question regarding macros as I see [include ./cfgs/misc-macros.cfg] points to that file. I am assuming adding additional macros to misc-macros.cfg will be sufficient.

**Comment by bassamanator** [comment]
This is probably your [solution](https://www.klipper3d.org/G-Codes.html#firmware_retraction): `firmware_retraction`. You probably want to `SET_RETRACTION` length and speed, along with temps. So those are your g-codes for when the material change occurs.

**Comment by Kalorado303** [comment]
Sorry I don't think I'm being clear as I am commenting here as I am going along.

I am having an issue when I call for next extruder in super slicer. If I don't add M600 to the tool change g-code in the slicer the print will error out in the beginning with an extruder0/1 not recognized error. When I add M600 in the custom g-code it works fine but pauses after my purge line and before my print starts so I have to start the print twice. 


Here is an example of a failed/error print g code w/o M600. The error says, The value "extruder 1" is not valid for EXTRUDER. 
M73 P0 R1
;TYPE:Custom
M104 S0 ; Stops PS/SS from sending temp waits separately
M140 S0
PRINT_START BED=70 HOTEND=235 CHAMBER=30
PURGE_LINE

;; If you are passing variables to PRINT_START, comment out above line and uncomment the last 3 lines instead. (I already did this)
;; Guide: https://github.com/AndrewEllis93/Print-Tuning-Guide/blob/main/articles/passing_slicer_variables.md
;; Make sure the variable names match (for example if you use "EXTRUDER" instead of "HOTEND")
M107
G21 ; set units to millimeters
G90 ; use absolute coordinates
M83 ; use relative distances for extrusion
ACTIVATE_EXTRUDER EXTRUDER=extruder1
;_TOOLCHANGE 1
M104 S235 ; set temperature
M107

Here is an example of a print that works but gets paused right after the purge line due to M600 in the slicer tool change custom g code. Meaning I have to resume the print before it starts.

M73 P0 R0
;TYPE:Custom
M104 S0 ; Stops PS/SS from sending temp waits separately
M140 S0
PRINT_START BED=70 HOTEND=235 CHAMBER=30
PURGE_LINE

;; If you are passing variables to PRINT_START, comment out above line and uncomment the last 3 lines instead. (I already did this)
;; Guide: https://github.com/AndrewEllis93/Print-Tuning-Guide/blob/main/articles/passing_slicer_variables.md
;; Make sure the variable names match (for example if you use "EXTRUDER" instead of "HOTEND")
M107
G21 ; set units to millimeters
G90 ; use absolute coordinates
M83 ; use relative distances for extrusion
M600
;_TOOLCHANGE 1
M104 S235 ; set temperature
M107

Any idea how to correct this?

Also can you tell me how to fix the resume button on the LCD menu under "octoprint" and perhaps rename that path to "Klipper" since were not running octoprint?

**Reply by bassamanator** [reply]
Ask about this in the SS repo, or even the PS repo, I'm no expert in these matters.

Moonraker makes use of the Octoprint api, feel free to remove `[octoprint_compat]` from `moonraker.conf`, it probably won't affect functionality. The resume function likely isn't working because the hotend is turned off, and again, this is a moonraker issue.

**Reply by Kalorado303** [reply]
Great will do. Thank you so much for your help!

**Reply by bassamanator** [reply]
You're welcome.

If you can figure this out, please post here and I'll see if I can add it to the repo somehow. This way the next person looking for this will have an easier time.
